torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -795,13 +823,31 @@ class ShapePropagator {
     //   device         : always matching and preserved
     //   tensor inputs  : *
     //   tensor outputs : 1
-    static const register_formula_for broadcasting_ops{
+    static const register_formula_for broadcasting_ops_arithmetic{
         {
             // Tensor-Tensor operators
             "aten::add(Tensor self, Tensor other, *, Scalar alpha) -> Tensor",
             "aten::sub(Tensor self, Tensor other, *, Scalar alpha) -> Tensor",
             "aten::mul(Tensor self, Tensor other) -> Tensor",
             "aten::div(Tensor self, Tensor other) -> Tensor",
+        },
+        [this](Node* node) -> type_vec_t {
+          if (auto maybe_tensor_types =
+                  gatherTensorTypes<DimensionedTensorType>(node)) {
+            AT_ASSERT(maybe_tensor_types->size() >= 2);
+            return {broadcast(*maybe_tensor_types, getScalarType(node))};
```

Comment:
Since the output type is different if `alpha` is an int than if `alpha` is a float, don't we have to give up on analysis if we find that the `alpha` type is a scalar input ?

Reply:
ah I think I currently have a bug in the handling of alpha elsewhere (in that I'm not considering the dtype of alpha). I'll look into this tomorrow and figure out what the correct behavior should be. it may be that I just want to error if alpha is a float when the tensors are ints.

Reply:
gotcha. shape analysis shouldn't error if it sees something that will error in the runtime because for example that op may not be run. in that case you can just give up on setting the output type

Reply:
I'm not seeing docs for alpha in the add() function, but under baddbmm I see "For inputs of type FloatTensor or DoubleTensor, arguments beta and alpha must be real numbers, otherwise they should be integers."
I'm going to assume the same applies here and error on that case.

Does anything actually need to change in shape analysis once I fix this in the op? Since that combination of parameters would fail anyway I assume this code would just never get reached under that condition?

Reply:
(this is on your branch)
```
print(torch.tensor(1).add(torch.tensor(2).dtype)
torch.int64
print(torch.tensor(0).add(torch.tensor(1), .5).dtype)
torch.float64
```

I don't think you're accounting for alpha in setting the output type. If both tensors aren't floats and the concrete type of the ScalarType input isn't known you don't have enough information to set the output type.

Reply:
ah I seem to have missed some alternate dispatch I'll need to find. It works as expected for kwargs but seems to behave differently for the positional args:
```
>>> print(torch.ones(1, dtype=torch.int).add(1, alpha=.5).dtype)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: For integral input tensors, argument alpha must not be a floating point number.
```

Reply:
With help from @colesbury found that this is actually calling the deprecated function `add(Tensor self, Scalar alpha, Tensor other)`. Since `other` is actually the third argument in this call, the float64 dtype is correct. If you actually pass ints for self and other, and pass a float for `alpha`, you'll get an error.